drivers/iio/adc/axi_pulse_capture.c

```
@@ -0,0 +1,453 @@
+			if (pulse->saved_en == LASER_ENABLE) {
+				pulse_capture_iowrite(pulse,
+					ADI_REG_CONFIG, PULSE_GEN_LOAD);
+				pulse_capture_iowrite(pulse,
+					ADI_REG_DRIVER_ENABLE, LASER_ENABLE);
+			}
+			/* clear the irq */
+			pulse_capture_iowrite(pulse,
+				ADI_REG_IRQ_SRC, ADI_IRQ_SRC_OTW_EXIT);
+			pulse->otw = false;
+		}
+	} else {
+		/* we are waiting for an enter transition */
+		if (irq_pending & ADI_IRQ_SRC_OTW_ENTER) {
+			/* save the current enable state */
+			pulse->saved_en = pulse_capture_ioread(pulse,
+						ADI_REG_DRIVER_ENABLE);
+			pulse_capture_iowrite(pulse,
+				ADI_REG_CONFIG, PULSE_GEN_RESET);
+			pulse_capture_iowrite(pulse,
+				ADI_REG_DRIVER_ENABLE, LASER_DISABLE);
+			/* clear the irq */
+			pulse_capture_iowrite(pulse,
+				ADI_REG_IRQ_SRC, ADI_IRQ_SRC_OTW_ENTER);
+			pulse->otw = true;
+		}
+	}
+	mutex_unlock(&indio_dev->mlock);
+
+	return IRQ_HANDLED;
+}
+
+static void axi_pulse_capture_dt_config(struct device *dev)
+{
+	struct device_node *np = dev->of_node;
+	bool disabled = false;
+	struct platform_device *pdev =
+			container_of(dev, struct platform_device, dev);
+	struct iio_dev *indio_dev = platform_get_drvdata(pdev);
+	struct axi_pulse_capture *pulse = iio_priv(indio_dev);
+
+	/* All interrupts are disabled on the core at startup.
+	 * Let's only enable the otw irqs.
+	 */
+	pulse_capture_iowrite(pulse, ADI_REG_IRQ_MASK,
+			  (ADI_IRQ_MASK_OUT_ALL &
+			  ~(ADI_IRQ_SRC_OTW_ENTER | ADI_IRQ_SRC_OTW_EXIT)));
+	/* Let's check if the the driver is disabled.
+	 * If it's not, we will enable it here according with the values
+	 * on the dts...
+	 */
+	disabled = of_property_read_bool(np, "adi,laser-driver-disabled");
+	dev_dbg(dev, "%s laser driver...",
+		(disabled == true) ? "Disable" : "Enable");
+
+	if (!disabled) {
+		u32 freq = 0, pulser_w = 0;
+		int ret = 0;
+
+		ret = of_property_read_u32(np, "adi,pulse-frequency", &freq);
+		if (ret < 0) {
+			dev_warn(dev, "Could not get pulse-frequency (%d)\n",
+				 ret);
+			return;
```

`return ret; `?

---

This API is not returning (void). And the idea is to have some default values which come from dts. So, I don't think we should fail the `probe()` just because we fail to do the default configurations... Should we actually return the error and fail the `probe()` (seemed a little extreme to me)?

---

Ok, then these properties should be optional. In the dt-binding they are stated as "Required properties".

---

Ok, it makes sense. I will also update that

---

Now, if one of these properties is not set, the driver won't enable the device by default. It can still be done via sysfs/iio. Hmm but i guess it makes sense to have some default values in case these (or one them) are not set.